Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,318 @@
+
+* `TensorProduct.lTensor_exact` says that one can tensor a short exact sequence on the left
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map g g'`
+in the presence of two short exact sequences.
+
+The proofs are those of Bourbaki, Algèbre, chap. 2, §3, n°6
+
+## TODO
+
+Add
+
+* Special case for short exact sequences of the form `M → N → N/M → 0`
+
+* Analogue for morphisms of algebras
+
+* Treat the noncommutative case
+
+* Treat the case of semirings (need to know what a short exact sequence is)
+
+-/
+
+
+namespace AddMonoidHom
+
+def Exact {M N P : Type _} [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  (f : outParam (M →+ N)) (g : outParam (N →+ P)) :=
+  g.ker = f.range
+
+end AddMonoidHom
+
+namespace LinearMap
+
+open AddMonoidHom LinearMap
+
+variable (R : Type _) [CommRing R]
+variable {M N P : Type _} [AddCommGroup M] [AddCommGroup N][AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+variable (f : M →ₗ[R] N) (g : N →ₗ[R] P)
+
+variable {R}
+def Exact := LinearMap.range f = LinearMap.ker g
+
+namespace Exact
+
+variable {f g}
+lemma comp_eq_zero (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [coe_comp, Function.comp_apply, zero_apply]
+  rw [← mem_ker, ← h]
+  apply mem_range_self
+
+section rTensor
+
+
+open scoped TensorProduct
+
+example (N' : Submodule R N) : N' →ₗ[R] N := by
+  exact Submodule.subtype N'
+
+example (N' : Submodule R N) : Submodule R (N ⊗[R] P) :=
+  LinearMap.range (rTensorHom P (N'.subtype))
+
+
+variable (Q : Type _) [AddCommGroup Q] [Module R Q]
+
+variable (g)
+lemma rTensor.mem_range {p : P} {q : Q} (hp : p ∈ LinearMap.range g) :
+  p ⊗ₜ[R] q ∈  LinearMap.range (rTensor Q g) := by
+  simp only [mem_range] at hp ⊢
+  obtain ⟨n, rfl⟩ := hp
+  use n ⊗ₜ[R] q
+  simp only [rTensor_tmul]
+
+variable {g}
+
+variable (hg : Function.Surjective g) (hfg : Exact f g)
+
+theorem rTensor.Surjective : Function.Surjective (rTensor Q g) := by
+  intro z
+  induction z using TensorProduct.induction_on with
+  | C0 => use 0; rw [map_zero]
+  | C1 p q =>
+      obtain ⟨n, rfl⟩ := hg p
+      use n ⊗ₜ[R] q
+      simp only [rTensor_tmul]
+  | Cp x y hx hy =>
+      obtain ⟨x, rfl⟩ := hx
+      obtain ⟨y, rfl⟩ := hy
+      use x + y
+      rw [map_add]
+
+/-
+private noncomputable
+def inverse : P →ₗ[R] (N ⧸ (LinearMap.range f)) := {
+  toFun := fun z ↦ Submodule.Quotient.mk ((hg z).choose)
+  map_add' := fun z z' ↦ by
+    dsimp
+    rw [← Submodule.Quotient.mk_add, Submodule.Quotient.eq, hfg, mem_ker]
+    simp only [map_sub, map_add, Exists.choose_spec (hg _), sub_self]
+  map_smul' := fun r z ↦ by
+    dsimp
+    rw [← Submodule.Quotient.mk_smul, Submodule.Quotient.eq, hfg, mem_ker]
+    simp only [map_sub, map_smul, Exists.choose_spec (hg _), sub_self] } -/
+
+private noncomputable
+def rTensor.inverse : P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ (LinearMap.range (rTensor Q f)) := by
+  apply TensorProduct.lift
+  exact {
+    toFun := fun z ↦ {
```

Comment:
`LinearMap.mk\2` can make this a bit shorter

Reply:
Indeed, thanks!